linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/Emitter.scala

```diff
@@ -695,6 +716,57 @@ final class Emitter(config: Emitter.Config) {
     }
   }
 
+  private class FullClassCache extends knowledgeGuardian.KnowledgeAccessor {
+    private[this] var _tree: WithGlobals[js.Tree] = null
+    private[this] var _lastUseESClass: Boolean = false
+    private[this] var _lastCtor: WithGlobals[js.Tree] = null
+    private[this] var _lastMemberMethods: List[WithGlobals[js.MethodDef]] = null
+    private[this] var _lastExportedMembers: List[WithGlobals[js.Tree]] = null
+    private[this] var _cacheUsed = false
+
+    override def invalidate(): Unit = {
+      super.invalidate()
+      _tree = null
+      _lastCtor = null
+      _lastMemberMethods = null
+      _lastExportedMembers = null
+    }
+
+    def startRun(): Unit = _cacheUsed = false
+
+    def getOrElseUpdate(useESClass: Boolean, ctor: WithGlobals[js.Tree],
+        memberMethods: List[WithGlobals[js.MethodDef]], exportedMembers: List[WithGlobals[js.Tree]],
+        compute: => WithGlobals[js.Tree]): WithGlobals[js.Tree] = {
+
+      @tailrec
+      def allSame[A <: AnyRef](xs: List[A], ys: List[A]): Boolean = {
+        xs.isEmpty == ys.isEmpty && {
+          xs.isEmpty ||
+          ((xs.head eq ys.head) && allSame(xs.tail, ys.tail))
+        }
+      }
+
+      if (_tree == null || (_lastCtor ne ctor) || !allSame(_lastMemberMethods, memberMethods) ||
+          !allSame(_lastExportedMembers, exportedMembers)) {
```

Comment:
I'm not sure where the right place to put this comment is:

I feel like this needs some smoke tests, because we start relying on reference equality. While I don't think this is a problem, it is a very non-local property, so IMO it's very easy we break caching without noticing it.

So I feel we need a test to make sure we actually get cache hits here (unclear to me how exactly this test would look though).

Reply:
I added a new linker test `BasicLinkerBackendTest` that checks that a no-op second run does not invalidate anything (except the one module initializer tree). If the `eq`s here fail, that will be reflected in what the `BasicLinkerBackend` recomputes and therefore its stats logging.